lib/stm32/h7/rcc.c

```diff
@@ -6,16 +6,333 @@
  *
  * LGPL License Terms @ref lgpl_license
  */
-
+#include <stddef.h>
 #include <libopencm3/cm3/assert.h>
 #include <libopencm3/stm32/rcc.h>
 #include <libopencm3/stm32/pwr.h>
 #include <libopencm3/stm32/flash.h>
 
-/**@{*/
+#define PLL1_INDEX			0
+#define PLL2_INDEX			1
+#define PLL3_INDEX			2
+#define HZ_PER_MHZ		 	1000000UL
+#define HZ_PER_KHZ			1000UL
+
+/* Local private copy of the clock configuration for providing user with clock tree data. */
+static struct {
+	uint16_t sysclk_mhz;
+	uint16_t cpu_mhz;
+	uint16_t hclk_mhz;
+	struct {
+		uint16_t pclk1_mhz;		/* APB1 clock. */
+		uint16_t pclk2_mhz;		/* APB2 clock. */
+		uint16_t pclk3_mhz;		/* APB3 clock. */
+		uint16_t pclk4_mhz;		/* APB4 clock. */
+	} per;
+	struct {					/* Each PLL output set of data. */
+		uint16_t p_mhz;
+		uint16_t q_mhz;
+		uint16_t r_mhz;
+	} pll[3];
```

Comment:
why go to all the effort of structs and then force people to use magic indexes?  Either don't nest them, or name them so they're obvious!

Reply:
It's a struct and an array so that it can be easily handled in a loop. The reason #define indexing is for documentation purposes only, because loop from 0-2 when the docs state PLL1-PLL3 is a bit confusing, so the macros are to document this.  I don't know a good way around ST's insistence on one-based numbering on a system that will be coded in a zero-based language. (although, we have found some of their peripherals actually have zero based defs too) 

I could have a dummy pll[0] that is unused?

Reply:
Okay...I refactored this to be explicit values, and unrolled the loop into a function call. It's much cleaner now with the register defs and indexing.